debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/AbstractTransactionCachingLogMinerEventProcessor.java

```diff
@@ -0,0 +1,435 @@
+        removeEventsWithTransaction(transaction);
+    }
+
+    @Override
+    protected void finalizeTransactionCommit(String transactionId, Scn commitScn) {
+        getAbandonedTransactionsCache().remove(transactionId);
+        // cache recently committed transactions by transaction id
+        if (getConfig().isLobEnabled()) {
+            getProcessedTransactionsCache().put(transactionId, commitScn.toString());
+        }
+    }
+
+    @Override
+    protected void finalizeTransactionRollback(String transactionId, Scn rollbackScn) {
+        final T transaction = getTransactionCache().get(transactionId);
+        if (transaction != null) {
+            removeEventsWithTransaction(transaction);
+            getTransactionCache().remove(transactionId);
+        }
+        getAbandonedTransactionsCache().remove(transactionId);
+        if (getConfig().isLobEnabled()) {
+            getProcessedTransactionsCache().put(transactionId, rollbackScn.toString());
+        }
+    }
+
+    @Override
+    protected void resetTransactionToStart(T transaction) {
+        super.resetTransactionToStart(transaction);
+        // Flush the change created by the super class to the transaction cache
+        getTransactionCache().put(transaction.getTransactionId(), transaction);
+    }
+
+    @Override
+    protected void handleSchemaChange(LogMinerEventRow row) throws InterruptedException {
+        super.handleSchemaChange(row);
+        if (row.getTableName() != null) {
+            getSchemaChangesCache().put(row.getScn().toString(), row.getTableId().identifier());
+        }
+    }
+
+    @Override
+    protected void addToTransaction(String transactionId, LogMinerEventRow row, Supplier<LogMinerEvent> eventSupplier) {
+        if (getAbandonedTransactionsCache().contains(transactionId)) {
+            LOGGER.warn("Event for abandoned transaction {}, skipped.", transactionId);
+            return;
+        }
+        if (!isRecentlyProcessed(transactionId)) {
+            T transaction = getTransactionCache().get(transactionId);
+            if (transaction == null) {
+                LOGGER.trace("Transaction {} is not in cache, creating.", transactionId);
+                transaction = createTransaction(row);
+            }
+
+            if (isTransactionOverEventThreshold(transaction)) {
+                abandonTransactionOverEventThreshold(transaction);
+                return;
+            }
+
+            String eventKey = transaction.getEventId(transaction.getNextEventId());
+            if (!getEventCache().containsKey(eventKey)) {
+                // Add new event at eventId offset
+                LOGGER.trace("Transaction {}, adding event reference at key {}", transactionId, eventKey);
+                getEventCache().put(eventKey, eventSupplier.get());
+                metrics.calculateLagFromSource(row.getChangeTime());
+                inMemoryPendingTransactionsCache.putOrIncrement(transaction.getTransactionId());
```

Comment:
So I get why this pending transaction cache exists, you split the `MemoryTransaction` to where events are no longer stored in the Transaction like we did for Infinispan, which means we no longer have a solid way to get the event cache size per transaction.  But do we know if this introduces any measurable overhead for heap-based implementations, particularly when large transactions are at play?

Reply:
I also failed to mention, the `Events` array that is maintained in the `MemoryTransaction`, it would seem that tis no longer applicable, so it can be removed, yes?

Reply:
I'm not sure what overhead this will introduce.  If we remove the events from the MemoryTransaction, seems like things might balance out.  Would need to run profiling against this to know.

Reply:
Well, the `MemoryTransaction's event array is no longer used; it's just an empty collection that serves no purpose; hence, my request to remove it.  

However, my concern is the difference between a `List#size` and `List#add`; we have this repeated map put/get+increment tuple. I suspect the latter loses here in performance.

Reply:
The in memory impl kept the events as part of the transaction whereas infinispan used this in memory pending cache.  Should the in memory pending cache be turned into a transient count on the transaction itself?